docs/architecture/adr-038-state-listening.md

```diff
@@ -204,60 +204,158 @@ func (rs *Store) CacheMultiStore() types.CacheMultiStore {
-    ListenDeliverTx(ctx types.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) error
-    // HaltAppOnDeliveryError whether or not to halt the application when delivery of massages fails
-    // in ListenBeginBlock, ListenEndBlock, ListenDeliverTx. When `false, the app will operate in fire-and-forget mode.
-    // When `true`, the app will gracefully halt and stop the running node. Uncommitted blocks will
-    // be replayed to all listeners when the node restarts and all successful listeners that received data
-    // prior to the halt will receive duplicate data. Whether or not a listener operates in a fire-and-forget mode
-    // is determined by the listener's configuration property `halt_app_on_delivery_error = true|false`.
-    HaltAppOnDeliveryError() bool
+	// ListenBeginBlock updates the streaming service with the latest BeginBlock messages
+	ListenBeginBlock(blockHeight int64, req []byte, res []byte) error
+	// ListenEndBlock updates the steaming service with the latest EndBlock messages
+	ListenEndBlock(blockHeight int64, req []byte, res []byte) error
+	// ListenDeliverTx updates the steaming service with the latest DeliverTx messages
+	ListenDeliverTx(blockHeight int64, req []byte, res []byte) error
+	// ListenStoreKVPair updates the steaming service with the latest StoreKVPair messages
+	ListenStoreKVPair(blockHeight int64, data []byte) error
 }
 
-// StreamingService interface for registering WriteListeners with the BaseApp and updating the service with the ABCI messages using the hooks
-type StreamingService interface {
-    // Stream is the streaming service loop, awaits kv pairs and writes them to a destination stream or file
-    Stream(wg *sync.WaitGroup) error
-    // Listeners returns the streaming service's listeners for the BaseApp to register
-    Listeners() map[types.StoreKey][]store.WriteListener
-    // ABCIListener interface for hooking into the ABCI messages from inside the BaseApp
-    ABCIListener
-    // Closer interface
-    io.Closer
+// StreamingService struct for registering WriteListeners with the BaseApp and updating the service with the ABCI messages using the hooks
+type StreamingService struct {
+	// Listeners returns the streaming service's listeners for the BaseApp to register
+	Listeners() map[types.StoreKey][]store.WriteListener
+	// ABCIListener interface for hooking into the ABCI messages from inside the BaseApp
+	ABCIListener ABCIListener
+	// StopNodeOnErr stops the node when true
+	StopNodeOnErr bool
 }
 ```
 
-#### BaseApp registration
+We will also introduce a new `KVStoreListener` struct as the writer for use in `StoreKVPairWriteListener`. `KVStoreListener` will expose `ABCIListener` for streaming state change events via `ABCIListener.ListenStoreKVPair`, a `BlockHeight` func for determining the block height for which the state change events belong to and a boolean flag for determine whether the node needs to be stopped or not.
```

Comment:
As I read this `KVStoreListener` clobbers too much with `ListenKVStorePair` and I find myself having to continuously go back and understand what these things are.

I highly recommend renaming `KVStoreListener` to `KVStoreWriter` as that is more clear and since that it's actually what it is.